src/metabase/query_processor/card.clj

```diff
@@ -191,33 +191,38 @@
       :or   {constraints (qp.constraints/default-query-constraints)
              context     :question
              qp-runner   qp/process-query-and-save-execution!}}]
   {:pre [(int? card-id) (u/maybe? sequential? parameters)]}
-  (let [run   (or run
-                  ;; param `run` can be used to control how the query is ran, e.g. if you need to
-                  ;; customize the `context` passed to the QP
-                  (^:once fn* [query info]
-                   (qp.streaming/streaming-response [{:keys [rff context]} export-format (u/slugify (:card-name info))]
-                     (qp-runner query info rff context))))
-        card  (api/read-check (t2/select-one [Card :id :name :dataset_query :database_id :cache_ttl :collection_id
-                                              :dataset :result_metadata :visualization_settings]
-                                             :id card-id))
-        query (-> (assoc (query-for-card card parameters constraints middleware {:dashboard-id dashboard-id}) :async? true)
-                  (update :middleware (fn [middleware]
-                                        (merge
-                                         {:js-int-to-string? true :ignore-cached-results? ignore_cache}
-                                         middleware))))
-        info  (cond-> {:executed-by            api/*current-user-id*
-                       :context                context
-                       :card-id                card-id
-                       :card-name              (:name card)
-                       :dashboard-id           dashboard-id
-                       :visualization-settings (:visualization_settings card)}
-                (and (:dataset card) (seq (:result_metadata card)))
-                (assoc :metadata/dataset-metadata (:result_metadata card)))]
+  (let [run       (or run
+                      ;; param `run` can be used to control how the query is ran, e.g. if you need to
+                      ;; customize the `context` passed to the QP
+                      (^:once fn* [query info]
+                       (qp.streaming/streaming-response [{:keys [rff context]} export-format (u/slugify (:card-name info))]
+                         (qp-runner query info rff context))))
+        dash-viz  (when (not= context :question)
+                    (t2/select-one-fn :visualization_settings :model/DashboardCard :id dashcard-id))
+        card      (api/read-check (t2/select-one [Card :id :name :dataset_query :database_id :cache_ttl :collection_id
+                                                  :dataset :result_metadata :visualization_settings]
+                                                 :id card-id))
+        query     (-> (query-for-card card parameters constraints middleware {:dashboard-id dashboard-id})
+                      (update :viz-settings (fn [viz] (when (or (seq viz) (seq dash-viz))
```

Comment:
The `query` that will be run by the query processor. Has a :viz-settings key which may be nil or may have the card's viz settings. Only update the key if either card or dashcard viz settings exist and aren't empty. This is because the private fn in this namespace `viz-settings` will first look for `:viz-settings` on the query. Only if that's nil will it instead look up the card's viz settings from the qp store. So, an empty map short circuits this logic and causes other tests to fail.

Reply:
Do you need the or logic in the update clause? I think you can safely update with merge as in the following examples:
```clojure
(update {} :viz-settings merge {:abc 123})
=> {:viz-settings {:abc 123}}
(update nil :viz-settings merge {:abc 123})
=> {:viz-settings {:abc 123}}
(update {:viz-settings {:abc "FOO"}} :viz-settings merge {:abc 123})
=> {:viz-settings {:abc 123}}
(update {} :viz-settings merge nil)
=> {:viz-settings nil}
(update {:viz-settings nil} :viz-settings merge nil)
=> {:viz-settings nil}
(update {:viz-settings {}} :viz-settings merge nil)
=> {:viz-settings {}}
```

Reply:
The reason I did this  was for the case where the card's viz settings were not nil but the dashcard viz settings were empty {} The logic in the middleware was such that if you pass an empty map in, it won't bother to look up card viz settings from the qp store. So, tests started failing. 

I've changed the logic in the middleware and removed the (or ...) here.